@ariakit/test: click/mouseDown/mouseUp still fire mousedown/mouseup after a canceled pointerdown, unlike real browsers

### Current behavior

Per the Pointer Events spec, calling `event.preventDefault()` on `pointerdown` sets the *PREVENT MOUSE EVENT* flag for that pointer: the compatibility mouse events `mousedown`/`mousemove`/`mouseup` are suppressed until the pointer is released, while boundary events and `click` still fire. Chrome, Firefox, Safari, and `@testing-library/user-event` (which parts of this package are modeled on, see the attribution at [dispatch.ts#L1](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/dispatch.ts#L1)) all implement this. Canceling `pointerdown` is the standard way for a component to opt out of the mouse-event sequence — most commonly to keep focus from moving on press.

`@ariakit/test` does not implement this suppression. In `mouseDown`, the `pointerdown` result is captured into `defaultAllowed`, but the `mousedown` dispatch is gated only on the element being enabled ([mouse-down.ts#L70-L81](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/mouse-down.ts#L70-L81)):

```ts
let defaultAllowed = await dispatch.pointerDown(element, options);

if (!disabled) {
  // Mouse events are not called on disabled elements
  if (!(await dispatch.mouseDown(element, { detail: 1, ...options }))) {
    defaultAllowed = false;
  }
}
```

`defaultAllowed` only gates the focus/selection block that follows ([mouse-down.ts#L79-L104](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/mouse-down.ts#L79-L104)) — so the *default action* (focus) is correctly skipped when `pointerdown` is canceled, but the `mousedown` event a real browser would never fire is still dispatched, and any `mousedown` listener on the component runs.

`mouseUp` has the same gap: it fires `mouseup` unconditionally after `pointerup`, with no knowledge that the matching `pointerdown` was canceled ([mouse-up.ts#L25-L30](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/mouse-up.ts#L25-L30)). `click()` strings these together without threading the cancellation through ([click.ts#L147-L185](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/click.ts#L147-L185)), even though its own JSDoc promises fidelity to the real event sequence ([click.ts#L126-L129](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/click.ts#L126-L129)). `mouseDown`, `mouseUp`, `click`, and `tap` are all public API ([index.ts#L8-L14](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/index.ts#L8-L14)), and `tap` inherits the gap through `click`.

This was verified against real Chromium with Playwright: clicking a button whose `pointerdown` handler calls `preventDefault()` fires neither `mousedown` nor `mouseup`, does not move focus, yet still fires `click`. A component that cancels `pointerdown` to opt out of mouse events will have its `mousedown`/`mouseup` handlers run under `@ariakit/test` even though they never run in a real browser — so tests can pass against behavior that is wrong in production, or fail against behavior that is correct.

### Steps to reproduce the bug

For a testing library, running a test is the consumer reproduction.

1. Open the React template: <https://stackblitz.com/github/ariakit/ariakit/tree/main/templates/react?file=src%2Fapp.tsx>. Replace `src/app.tsx` with the code below — a button that cancels `pointerdown` (the standard "prevent focus / mouse side-effects on press" pattern) and mirrors which compatibility mouse events reached it:

```tsx
import type {
  MouseEvent as ReactMouseEvent,
  PointerEvent as ReactPointerEvent,
} from "react";
import { useState } from "react";

export default function App() {
  const [mouseDownFired, setMouseDownFired] = useState(false);
  const [mouseUpFired, setMouseUpFired] = useState(false);
  const [clickFired, setClickFired] = useState(false);
  const [focused, setFocused] = useState(false);

  function handlePointerDown(event: ReactPointerEvent<HTMLButtonElement>) {
    // Opt out of the compatibility mouse events (and focus-on-press).
    event.preventDefault();
  }

  return (
    <div>
      <button
        type="button"
        onPointerDown={handlePointerDown}
        onMouseDown={() => setMouseDownFired(true)}
        onMouseUp={() => setMouseUpFired(true)}
        onClick={() => setClickFired(true)}
        onFocus={() => setFocused(true)}
        onBlur={() => setFocused(false)}
      >
        Press me
      </button>
      <p>
        <output aria-label="mousedown">{String(mouseDownFired)}</output>{" "}
        <output aria-label="mouseup">{String(mouseUpFired)}</output>{" "}
        <output aria-label="click">{String(clickFired)}</output>{" "}
        <output aria-label="focused">{String(focused)}</output>
      </p>
    </div>
  );
}
```

2. Add a test (e.g. `src/test.ts`) and run it with `@ariakit/test` + happy-dom:

```ts
import { click, q } from "@ariakit/test";
import { expect, test } from "vitest";

function getStatusText(name: string) {
  return q.status.ensure(name).textContent;
}

test("canceling pointerdown suppresses the compatibility mouse events", async () => {
  await click(q.button.ensure("Press me"));

  // Real browsers fire neither mousedown nor mouseup when pointerdown is
  // canceled, and do not move focus, while click still fires.
  expect(getStatusText("mousedown")).toBe("false");
  expect(getStatusText("mouseup")).toBe("false");
  expect(getStatusText("click")).toBe("true");
  expect(getStatusText("focused")).toBe("false");
});
```

3. Run `pnpm test-react test-pointerdown-mouse-suppress` (in this repo, the sandbox at `app/src/sandbox/test-pointerdown-mouse-suppress/`). The test fails on the very first assertion:

```
AssertionError: expected 'true' to be 'false'   // mousedown fired despite the canceled pointerdown
```

The same user scenario was verified in a real browser with Playwright (`app/src/sandbox/test-pointerdown-mouse-suppress/test-browser.ts`, a real mouse click on the rendered button): it passes in Chromium, where `mousedown`/`mouseup` are suppressed and `click` still fires.

### Expected behavior

`click`/`mouseDown`/`mouseUp`/`tap` should mirror the Pointer Events spec's PREVENT MOUSE EVENT behavior: when a `pointerdown` listener calls `preventDefault()`, the compatibility `mousedown` and `mouseup` events are suppressed for that interaction (no `mousedown`/`mouseup` listeners run, no focus moves), while `click` still fires. A subsequent un-canceled press should fire the full sequence again. That is what real browsers, and `@testing-library/user-event`, do.

### Workaround

Consumer-side, in test code only. Instead of the high-level `click()` (which always fires the compatibility mouse events), replicate the press sequence by hand and gate the mouse events on `dispatch.pointerDown`'s return value, which is `false` when `preventDefault()` was called:

```ts
import { dispatch, hover, q } from "@ariakit/test";

async function clickWithMouseSuppression(element: Element) {
  await hover(element);
  const pointerAllowed = await dispatch.pointerDown(element, { detail: 1 });
  if (pointerAllowed) {
    await dispatch.mouseDown(element, { detail: 1 });
  }
  await dispatch.pointerUp(element, { detail: 1 });
  if (pointerAllowed) {
    await dispatch.mouseUp(element, { detail: 1 });
  }
  // click still fires regardless of the pointerdown cancellation.
  await dispatch.click(element, { detail: 1 });
}

await clickWithMouseSuppression(q.button.ensure("Press me"));
```

Verified: with this in place the test above passes on the current `@ariakit/test`. (This is a faithful but partial stand-in for `click()` — it omits the focus/selection default-action and visible-parent walk — so it's a workaround for the specific cancel-pointerdown case, not a general `click` replacement.)

### Possible solutions

Track the cancellation per-document — mirroring the spec's PREVENT MOUSE EVENT flag — and suppress the compatibility mouse events. The state must live per-document (not per-element or in a local), because `mouseDown` and `mouseUp` are independent public functions and `click()` may walk to a visible parent between them; this follows the established per-document `lastHovered` expando in [hover.ts#L6-L8](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/hover.ts#L6-L8). Derived from the source (not test-suite-verified here, per the parallel-safety constraint on `packages/`):

Define the shared type once in `__utils.ts`:

```diff
 export type TextField = HTMLInputElement | HTMLTextAreaElement;
+
+// Per-document pointer state mirroring the Pointer Events spec's PREVENT MOUSE
+// EVENT flag: set when pointerdown is canceled so the compatibility mouse
+// events (mousedown, mouseup) are suppressed until the pointer is released.
+export type DocumentWithPointerState = Document & {
+  pointerDownPrevented?: boolean;
+};
```

In `mouseDown`, record the flag and gate the `mousedown` dispatch on the `pointerdown` result (`getDocument` is already imported here):

```diff
-    let defaultAllowed = await dispatch.pointerDown(element, options);
+    const pointerDefaultAllowed = await dispatch.pointerDown(element, options);
+
+    // Canceling pointerdown sets the spec's PREVENT MOUSE EVENT flag: the
+    // compatibility mouse events are suppressed until the pointer is released,
+    // while click still fires. Store it on the document so mouseUp can read it.
+    (getDocument(element) as DocumentWithPointerState).pointerDownPrevented =
+      !pointerDefaultAllowed;
+
+    let defaultAllowed = pointerDefaultAllowed;

-    if (!disabled) {
+    if (!disabled && pointerDefaultAllowed) {
       // Mouse events are not called on disabled elements
       if (!(await dispatch.mouseDown(element, { detail: 1, ...options }))) {
         defaultAllowed = false;
       }
     }
```

This preserves the existing focus/selection block: it was already skipped when `pointerdown` was canceled (since `defaultAllowed` started `false`), which is spec-correct because that default belongs to the now-suppressed `mousedown`. The disabled-element focus-fallback path is unchanged.

In `mouseUp`, read and clear the flag on every `pointerup` — before the disabled early return, since the suppression window ends on release regardless of the element's state:

```diff
     await dispatch.pointerUp(element, options);

+    // The PREVENT MOUSE EVENT suppression lasts until the pointer is released,
+    // so read and clear the flag on every pointerup.
+    const document = element.ownerDocument as DocumentWithPointerState;
+    const pointerDownPrevented = document.pointerDownPrevented;
+    document.pointerDownPrevented = false;
+
     // mouseup is not called on disabled elements
     if (disabled) return;

+    // The matching pointerdown was canceled, suppressing the compatibility
+    // mouseup as well.
+    if (pointerDownPrevented) return;
+
     await dispatch.mouseUp(element, { detail: 1, ...options });
```

`click()` needs no change: it keeps dispatching `click` after `mouseUp` ([click.ts#L168-L182](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/click.ts#L168-L182)), which already matches browsers — canceling `pointerdown` suppresses only the compatibility mouse events, not `click`. `tap()` inherits the fix through `click()`. The flag is overwritten on every new press, so it cannot leak stale state between interactions. (The spec also suppresses compatibility `mousemove`; gating `hover()`'s `mouseMove` on the same flag is a straightforward follow-up if drag-style sequences ever need it — `click`/`tap` never hover mid-press.) Since `@ariakit/test` is a published package and this changes observable event sequencing, a patch changeset is warranted.

*Found by an automated multi-agent audit of the repo at commit [`11bd4458d`](https://github.com/ariakit/ariakit/tree/11bd4458de6447b996dd882ea0c85eb801a6cd03) and verified with a real-browser Playwright reproduction (user-level interactions only).*

--- a/.changeset/300-test-pointerdown-mouse-suppression.md
+++ b/.changeset/300-test-pointerdown-mouse-suppression.md
@@ -1,0 +1,5 @@
+---
+"@ariakit/test": patch
+---
+
+Fixed `click`, `tap`, `mouseDown`, and `mouseUp` to suppress compatibility mouse events after a canceled `pointerdown`.

--- a/packages/ariakit-test/src/__utils.ts
+++ b/packages/ariakit-test/src/__utils.ts
@@ -4,6 +4,8 @@ export type DirtiableElement = Element & { dirty?: boolean };
 
 export type TextField = HTMLInputElement | HTMLTextAreaElement;
 
+const preventMouseEvents = new WeakMap<Document, boolean>();
+
 export const isHappyDOM = typeof window !== "undefined" && "happyDOM" in window;
 
 export const isBrowser =
@@ -103,6 +105,14 @@ export function setActEnvironment(value: boolean) {
     scope.IS_REACT_ACT_ENVIRONMENT = previousValue;
   };
   return restoreActEnvironment;
+}
+
+export function getPreventMouseEvents(document: Document) {
+  return preventMouseEvents.get(document) ?? false;
+}
+
+export function setPreventMouseEvents(document: Document, value: boolean) {
+  preventMouseEvents.set(document, value);
 }
 
 let wrapAsyncDepth = 0;

--- a/packages/ariakit-test/src/mouse-down.test.ts
+++ b/packages/ariakit-test/src/mouse-down.test.ts
@@ -1,5 +1,7 @@
 import { beforeEach, expect, test } from "vitest";
+import { click } from "./click.ts";
 import { mouseDown } from "./mouse-down.ts";
+import { mouseUp } from "./mouse-up.ts";
 import { q } from "./query.ts";
 import { select } from "./select.ts";
 
@@ -96,4 +98,81 @@ test("mouseDown on a plain text target clears the document selection", async () 
   await selectParagraphText();
   await mouseDown(q.text("Plain text target"));
   expect(document.getSelection()?.toString()).toBe("");
+});
+
+test("click suppresses mouse events when pointerdown is prevented", async () => {
+  document.body.innerHTML = `
+    <input aria-label="Before" />
+    <button type="button">Press me</button>
+  `;
+
+  const input = q.textbox.ensure("Before");
+  const button = q.button.ensure("Press me");
+  const events: string[] = [];
+  let preventPointerDown = true;
+
+  input.focus();
+
+  button.addEventListener("pointerdown", (event) => {
+    events.push(event.type);
+    if (preventPointerDown) {
+      event.preventDefault();
+    }
+  });
+  button.addEventListener("mousedown", (event) => events.push(event.type));
+  button.addEventListener("focus", (event) => events.push(event.type));
+  button.addEventListener("pointerup", (event) => events.push(event.type));
+  button.addEventListener("mouseup", (event) => events.push(event.type));
+  button.addEventListener("click", (event) => events.push(event.type));
+
+  await click(button);
+
+  expect(events).toEqual(["pointerdown", "pointerup", "click"]);
+  expect(input).toHaveFocus();
+
+  events.length = 0;
+  preventPointerDown = false;
+
+  await click(button);
+
+  expect(events).toEqual([
+    "pointerdown",
+    "mousedown",
+    "focus",
+    "pointerup",
+    "mouseup",
+    "click",
+  ]);
+  expect(button).toHaveFocus();
+});
+
+test("mouseUp suppresses mouseup after a prevented pointerdown", async () => {
+  document.body.innerHTML = `<button type="button">Press me</button>`;
+
+  const button = q.button.ensure("Press me");
+  const events: string[] = [];
+  let preventPointerDown = true;
+
+  button.addEventListener("pointerdown", (event) => {
+    events.push(event.type);
+    if (preventPointerDown) {
+      event.preventDefault();
+    }
+  });
+  button.addEventListener("mousedown", (event) => events.push(event.type));
+  button.addEventListener("pointerup", (event) => events.push(event.type));
+  button.addEventListener("mouseup", (event) => events.push(event.type));
+
+  await mouseDown(button);
+  await mouseUp(button);
+
+  expect(events).toEqual(["pointerdown", "pointerup"]);
+
+  events.length = 0;
+  preventPointerDown = false;
+
+  await mouseDown(button);
+  await mouseUp(button);
+
+  expect(events).toEqual(["pointerdown", "mousedown", "pointerup", "mouseup"]);
 });

--- a/packages/ariakit-test/src/mouse-down.ts
+++ b/packages/ariakit-test/src/mouse-down.ts
@@ -6,7 +6,7 @@ import {
   isTextbox,
   invariant,
 } from "@ariakit/utils";
-import { wrapAsync } from "./__utils.ts";
+import { setPreventMouseEvents, wrapAsync } from "./__utils.ts";
 import { blur } from "./blur.ts";
 import { dispatch } from "./dispatch.ts";
 import { focus } from "./focus.ts";
@@ -67,9 +67,12 @@ export function mouseDown(element: Element | null, options?: PointerEventInit) {
 
     const { disabled } = element as HTMLButtonElement;
 
-    let defaultAllowed = await dispatch.pointerDown(element, options);
+    const pointerDefaultAllowed = await dispatch.pointerDown(element, options);
+    setPreventMouseEvents(getDocument(element), !pointerDefaultAllowed);
 
-    if (!disabled) {
+    let defaultAllowed = pointerDefaultAllowed;
+
+    if (!disabled && pointerDefaultAllowed) {
       // Mouse events are not called on disabled elements
       if (!(await dispatch.mouseDown(element, { detail: 1, ...options }))) {
         defaultAllowed = false;

--- a/packages/ariakit-test/src/mouse-up.ts
+++ b/packages/ariakit-test/src/mouse-up.ts
@@ -1,5 +1,9 @@
-import { isVisible, invariant } from "@ariakit/utils";
-import { wrapAsync } from "./__utils.ts";
+import { getDocument, isVisible, invariant } from "@ariakit/utils";
+import {
+  getPreventMouseEvents,
+  setPreventMouseEvents,
+  wrapAsync,
+} from "./__utils.ts";
 import { dispatch } from "./dispatch.ts";
 
 /**
@@ -24,8 +28,14 @@ export function mouseUp(element: Element | null, options?: PointerEventInit) {
 
     await dispatch.pointerUp(element, options);
 
+    const document = getDocument(element);
+    const preventMouseEvents = getPreventMouseEvents(document);
+    setPreventMouseEvents(document, false);
+
     // mouseup is not called on disabled elements
     if (disabled) return;
+
+    if (preventMouseEvents) return;
 
     await dispatch.mouseUp(element, { detail: 1, ...options });
   });

